disperser/controller/encoding_manager.go

```diff
@@ -0,0 +1,200 @@
+		return nil, fmt.Errorf("NumRelayAssignment (%d) cannot be greater than NumRelays (%d)", config.NumRelayAssignment, len(config.AvailableRelays))
+	}
+	return &EncodingManager{
+		EncodingManagerConfig: config,
+		blobMetadataStore:     blobMetadataStore,
+		pool:                  pool,
+		encodingClient:        encodingClient,
+		chainReader:           chainReader,
+		logger:                logger.With("component", "EncodingManager"),
+
+		lastUpdatedAt: 0,
+	}, nil
+}
+
+func (e *EncodingManager) Start(ctx context.Context) error {
+	go func() {
+		ticker := time.NewTicker(e.PullInterval)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				err := e.HandleBatch(ctx)
+				if err != nil {
+					if errors.Is(err, errNoBlobsToEncode) {
+						e.logger.Warn("no blobs to encode")
+					} else {
+						e.logger.Error("failed to process a batch", "err", err)
+					}
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (e *EncodingManager) HandleBatch(ctx context.Context) error {
+	// Get a batch of blobs to encode
+	blobMetadatas, err := e.blobMetadataStore.GetBlobMetadataByStatus(ctx, v2.Queued, e.lastUpdatedAt)
+	if err != nil {
+		return err
+	}
+
+	if len(blobMetadatas) == 0 {
+		return errNoBlobsToEncode
+	}
+
+	currentBlockNumber, err := e.chainReader.GetCurrentBlockNumber(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get current block number: %w", err)
+	}
+
+	for _, blob := range blobMetadatas {
+		blobKey, err := blob.BlobHeader.BlobKey()
+		if err != nil {
+			e.logger.Error("failed to get blob key", "err", err, "requestedAt", blob.RequestedAt, "paymentMetadata", blob.BlobHeader.PaymentMetadata)
+			continue
+		}
+		e.lastUpdatedAt = blob.UpdatedAt
+
+		// Encode the blobs
+		e.pool.Submit(func() {
+			for i := 0; i < e.NumEncodingRetries+1; i++ {
+				encodingCtx, cancel := context.WithTimeout(ctx, e.EncodingRequestTimeout)
+				fragmentInfo, err := e.encodeBlob(encodingCtx, blobKey, blob)
+				cancel()
+				if err != nil {
+					e.logger.Error("failed to encode blob", "blobKey", blobKey.Hex(), "err", err)
+					continue
+				}
+				relayKeys, err := GetRelayKeys(e.NumRelayAssignment, e.AvailableRelays)
+				if err != nil {
+					e.logger.Error("failed to get relay keys", "err", err)
+					// Stop retrying
+					break
+				}
+				cert := &corev2.BlobCertificate{
+					BlobHeader:           blob.BlobHeader,
+					ReferenceBlockNumber: uint64(currentBlockNumber),
+					RelayKeys:            relayKeys,
+				}
+
+				storeCtx, cancel := context.WithTimeout(ctx, e.StoreTimeout)
+				err = e.blobMetadataStore.PutBlobCertificate(storeCtx, cert)
+				cancel()
+				if err != nil && !errors.Is(err, dispcommon.ErrAlreadyExists) {
+					e.logger.Error("failed to put blob certificate", "err", err)
+					continue
+				}
+
+				storeCtx, cancel = context.WithTimeout(ctx, e.StoreTimeout)
+				err = e.blobMetadataStore.MarkBlobEncoded(storeCtx, blobKey, fragmentInfo)
```

Comment:
Why are we updating the blob record with the `fragmentInfo` instead of just pushing it next to the `BlobCert`?

Reply:
It would be nicer to put it in BlobCert. I'm not sure if it belongs there though. Is it something that should go onchain?

Reply:
Couldn't we just store a wrapper at this stage which contains both the `BlobCert` and this additional metadata? I agree it should not be in the actual canonical `BlobCert` structure.

Reply:
My worry is that now there are two different data structures representing `BlobCert` (one used in disperser & relay, and another one in node). I'd rather keep one consistent struct across. Is updating the blob metadata that bad?

Reply:
Actually, ignore that. I ended up marshaling/unmarshaling them separately and putting them in BlobCertificate: https://github.com/Layr-Labs/eigenda/pull/846/commits/e76e3a62acddefcf9538aa731142d0fd8385f508